padinfo/boardgenerator.py

```diff
@@ -0,0 +1,69 @@
+class BoardGenerator(object):
+    """Generate a Dawnglare link with optional inversion."""
+
+    allowed_letters = "RGBLDHXJMPZ "
+    inversion = "-1"
+    dawnglare_link = "https://pad.dawnglare.com/?patt={}&showfill=1"
+
+    def __init__(self, board):
+        self.board = board.split()
+        self.board_height = None
+        self.board_width = None
+        self.invalid_board = False
+        self.invert = BoardGenerator.inversion in self.board
+        self.clean_board = self.get_clean_board()
+        self.invalid_orbs = not all(letter in BoardGenerator.allowed_letters for letter in self.clean_board)
+        self.board_length = len(self.clean_board)
+        self.eval_board_size()
+        self.link = self.get_link()
+
+    def get_clean_board(self):
+        board = self.board
+
+        if self.invert:
+            board.remove(BoardGenerator.inversion)
+            clean = ''.join(self.invert_fill(board))
+        else:
+            clean = ''.join(board)
+
+        return clean
+
+    def invert_fill(self, board):
+        inverted_board = []
+
+        try:
+            for i in range(len(board[0])):
+                build_string = ""
+                for j in range(len(board)):
+                    build_string += board[j][i]
+                inverted_board.append(build_string)
+        except:
+            self.invalid_board = True
+
+        return inverted_board
+
+    def eval_board_size(self):
+        length = self.board_length
+
+        if length == 20:
```

Comment:
again this is pretty optional, but this could be done via a class constant with a dictionary if you want. `board_sizes = { }` and I'll let you see if you can figure out the rest :)

You'd want to use the dict.get() method which returns `None` on a missing input so that you can set `invalid_board` in that case instead of getting a `KeyError`

(this is optional!!! it can also be a later PR!)

Reply:
I did this a little differently than you suggested but it seems clean with a lot less code.

Reply:
this is great!! the only thing is, in contrast to what I said about "avoiding local variables," I *personally* prefer to avoid dict calls inside of fstrings because having quotes inside of quotes kind of sucks (imo). Some people don't care about that and will do what you did. But I actually would say `height = self.board_sizes[self.board_length]['height']` and then sub `height` into the fstring now, and same thing for width. This is a bit personal-preference-y though.

Reply:
I reworked this one more time and took your advice on happy path and return ASAP.